app/src/main/java/com/nutomic/syncthingandroid/model/Completion.java

```diff
@@ -0,0 +1,87 @@
+package com.nutomic.syncthingandroid.model;
+
+import java.util.HashMap;
+
+/**
+ * This class caches remote folder and device synchronization
+ * completion indicators defined in {@link CompletionInfo#CompletionInfo}
+ * according to syncthing's REST "/completion" JSON result schema.
+ * Completion model of syncthing's web UI is completion[deviceId][folderId]
+ */
+public class Completion {
+
+    HashMap<String, HashMap<String, CompletionInfo>> deviceFolderMap =
+        new HashMap<String, HashMap<String, CompletionInfo>>();

```

Comment:
Perhaps something like this can reduce boilerplate code https://stackoverflow.com/a/7519422

Reply:
Didn't notice that before, it actually looks pretty complicated to have nested maps like this. Maybe you can add an inner class that holds the second map and simplifies things.

Reply:
Whoa... I'm quite out of ideas for today. I get your point but currently I haven't found an easier, shorter implementation yet. I also added two update functions with commit https://github.com/syncthing/syncthing-android/pull/1079/commits/a5b2c57cc725c4435c8bb38464fc9ef30fca0eeb that are called after a config reload took place to clean out folders and devices that were cached before in the Completion class and do no longer exist after the config reload. I verified this working "fastly" on a small setup. But the code is not optimized for a high count of folder and device pairs because searching through the hashmaps has nearly an O(n^2) complexity. I thought of replacing it with Collection.html#retainAll 
( http://download.oracle.com/javase/6/docs/api/java/util/Collection.html#retainAll%28java.util.Collection%29 ) but this has the worst complexity in doing single comparations between old and new config hierarchy. I would appreciate any help on this.